### DISCUSSION TITLE
Thumbnail Fixer, a tool to repair & enhance embedded thumbnails

### DISCUSSION THREAD

**Discussion by StefanOltmann** [libvips/libvips#4326] [Show and tell]
Hi,

Thank you again for your help and for creating this fantastic library!

I'm thrilled to share that I've just released a new tool called Thumbnail Fixer, which leverages libvips.

This tool repairs and enhances embedded thumbnails in JPG, PNG, and WebP files, improving their loading speed in many applications.

Feel free to check it out here:
https://github.com/stefanoltmann/thumbnail-fixer

Previously, I had developed a similar project privately, but its UI lacked polish, and it relied on SKIA, which was slower and more resource-intensive.

Now, with a refined UI and the power of libvips, I'm excited to share this project with the world. 😊

Kind regards and happy holidays,

Stefan

**Comment by jcupitt** [comment]
Oh, nice! It looks great @StefanOltmann.

You could consider adding support for AVIF and HEIC as well I guess. They support thumbnails, and they are often annoyingly out of sync with the main image.

**Reply by StefanOltmann** [reply]
That's a great idea, but due to the HEIC licensing issues, I'd prefer to steer clear of that topic for now. 😄 Perhaps one day, a brave forker will take it on! 😅

**Reply by jcupitt** [reply]
True! You could do AVIF though. 

I guess you've looked at parallel conversion too.

**Reply by StefanOltmann** [reply]
I'm not a lawyer, but I've heard there are concerns about AVIF due to its potential overlap with HEIC, particularly around licensing and patent issues. That seems too risky for me.

However, I'm looking into JXL, which I believe is the future of image formats, and it’s already supported by my metadata library. The prebuilt Windows binary I use from the official repository doesn't include JXL for some reason, so I had to leave it out in my first version.

In a later version I may rewrite the logic to process in parallel, yes. What I released now is just my MVP and I like to get those out early for feedback. 🙂

**Reply by StefanOltmann** [reply]
Ah, no JXL because it’s not in the web package which I choose for FFI. Understood.

https://github.com/libvips/build-win64-mxe/issues/40

**Reply by jcupitt** [reply]
You'd need the `all` version to get JXL compiled in. It's not in the `web` bundle because fuzzers are still finding bugs in libjxl every week or so.

For a local tool like this, where the user controls what files get fed in, I think the `all` package would be fine.

**Reply by jcupitt** [reply]
AVIF is fine, legally. It's only the codec in HEIC which is patented, not the file format, so if you use `av1` it's all ok.

**Reply by StefanOltmann** [reply]
> AVIF is fine, legally. It's only the codec in HEIC which is patented, not the file format, so if you use av1 it's all ok.

I appreciate the clarification. Back when I was researching this, I struggled to find reliable information. There was an extended discussion within the Exiv2 community about whether the tool should support reading and writing metadata for HEIF files.

This led to the following decision for my metadata library:
> In the processing of HEIC and AVIF files, we handle them as standard ISOBMFF-based files, adhering rigorously to the EIC/ISO 14496-12 specification. To preempt potential legal issues, we intentionally omit certain boxes outlined in the HEIC specification, notably the image size ("ispe") and image rotation ("irot") boxes. This approach extends to AVIF images, as they repurpose the same boxes.
( https://github.com/Ashampoo/kim?tab=readme-ov-file#regarding-heic--avif-metadata )

Are you suggesting that it’s permissible to reference the HEIF specification for information, given that the image container format itself isn’t patented? Is this definitively established, or is it more of an assumption? My metadata library strictly works with the container and does not perform any decoding or encoding of the actual image bytes.

If this understanding is correct, then using a version of libvips with AVIF support compiled in should be legally safe - as you said.

I am genuinely concerned that any misstep in this area could result in serious legal issues, potentially costing significant amounts of money and jeopardizing both my livelihood and my employer's stability.

**Reply by jcupitt** [reply]
@lovell can I ping you for a second opinion on this?

**Reply by jcupitt** [reply]
(Dirk, who wrote libheif, would be the authoritative expert, but I'm hesitant to bother him directly. You could maybe open an issue on libheif and ask for clarification, or perhaps look for past issues which have discussed this)

**Reply by lovell** [reply]
@StefanOltmann Does your question relate to a specific patent? Perhaps https://patents.google.com/patent/WO2016193543A1 held by Nokia?

**Reply by StefanOltmann** [reply]
I'm uncertain about the patent situation surrounding HEIC.

If I had a comprehensive list of patents related to the HEIC container format, I could determine whether creating new containers or modifying existing ones would comply with applicable laws. While it's well-known that HEIC is heavily patented, I currently lack an overview of the specific patents that might protect it.

In discussions around this topic in the Exiv2 issues, Jon Sneyers mentioned that it should be safe to use ISOBMFF as specified historically, as it’s old enough that any relevant patents are likely expired. This reasoning is why I added generic ISOBMFF support to my library, specifically for reading files.

For reference, here’s a detailed answer he gave:
https://github.com/libjxl/libjxl/issues/3236#issuecomment-1926861855

However, when it comes to manipulating or generating new HEIC files, I’m unsure if this would require a license.

Mainly because I would write boxes like "irot" and "ispe" that are defined in the HEIC specification. I don't worry about anything that was taken from the original ISOBMFF as is.

The key question is: Can we safely manipulate the container of a HEIC file, provided we don’t encode or decode the actual media bytes? Specifically, I’m curious if it’s permissible to include new thumbnail bytes in HEIC and AVIF files.

If working with HEIC containers is not legally safe, would AVIF be permissible? Both formats are similar in their use of the container, so clarity on this aspect is crucial.

**Reply by lovell** [reply]
The short/expensive answer is that you probably need a patent lawyer.

My non-lawyer interpretation of claims relating to HEIF the container is that they can mostly be ignored.

To pick an example, I attempted to search for patents that relate to HEIF the container where a claim is made on `irot`. The only patent I can find is https://patents.google.com/patent/US11985302B2 filed by Canon (and not Nokia) and the list of claims includes:

> 6. The method of claim 5, wherein the property box indicating the decoder configuration of the one or more images is a box with box type ‘hvcC’.

...which limits its application to HEVC-encoded images.

Perhaps it might be worth using https://patents.stackexchange.com/ to collate prior art on patent claims known to relate to HEIF the container.

(Please note that what is written above relates only to containers, and does not relate to codecs.)
